src/Integration.Vsix/Helpers/DocumentEvents/ActiveDocumentTracker.cs

```diff
@@ -0,0 +1,110 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace SonarLint.VisualStudio.Integration.Vsix.Helpers.DocumentEvents
+{
+    internal interface IActiveDocumentTracker : IDisposable
+    {
+        event EventHandler<DocumentFocusedEventArgs> OnDocumentFocused;
+    }
+
+    [Export(typeof(IActiveDocumentTracker))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class ActiveDocumentTracker : IActiveDocumentTracker, IVsSelectionEvents
+    {
+        private readonly ITextDocumentProvider textDocumentProvider;
+        private IVsMonitorSelection monitorSelection;
+        private uint cookie;
+        private bool disposed;
+
+        public event EventHandler<DocumentFocusedEventArgs> OnDocumentFocused;
+
+        [ImportingConstructor]
+        public ActiveDocumentTracker([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, ITextDocumentProvider textDocumentProvider)
+        {
+            this.textDocumentProvider = textDocumentProvider;
+
+            RunOnUIThread.Run(() =>
+            {
+                monitorSelection = serviceProvider.GetService<SVsShellMonitorSelection, IVsMonitorSelection>();
+
+                if (monitorSelection == null)
+                {
+                    throw new ArgumentNullException(nameof(monitorSelection));
+                }
+
+                monitorSelection.AdviseSelectionEvents(this, out cookie);
+            });
+        }
+
+        public int OnElementValueChanged(uint elementId, object oldValue, object newValue)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (newValue is IVsWindowFrame frame &&
+                IsFrameElement() && 
+                IsDocumentFrame())
+            {
+                var textDocument = textDocumentProvider.GetFromFrame(frame);
```

Comment:
`GetFromFrame` can return null, can't it?
Whose responsibility is it to handle this - this class (in which case it shouldn't raise the event in that case), or the event listeners (who would all then have to handle the document being null)?

If this class is making a guarantee that it will never raise with a null document then it's worth documenting in a doc comment on the event declaration.

Reply:
You're right; I'll add a null check + document that it shouldn't raise an event with a null document.